### PR TITLE
Basis support

### LINKED ISSUE #1735: Add support for basis compression
Add support for the basis 'super-compressed' texture format.

(See https://github.com/BinomialLLC/basis_universal).

The advantages to Playcanvas will be as follows:
- one single compressed texture for all platforms (ios, android, pc)
- final game package is smaller since we don't have to create multiple versions of a compressed texture
- the engine will easily support more runtime compressed formats thanks to the transcoder
- authoring time will be saved: the basis compression tool runs faster and better than the existing compressors (and compression is only required once, not multiple times for each target platform).

The possible disadvantages are:
- basis can't modify image dimensions at runtime, so PVR which has square-power-of-two restriction requires it's own file for non-square or non-power-of-two textures
- possible slight loss of image quality due to transcoding
- load times could be increased due to additional transcode step required at load time
- more involved runtime, basis transcoder requires wasm module loading

As Binomial continue to develop their compressor and transcoder the disadvantages should be improved.

### PR DESCRIPTION
**Pull request by slimbuck**
Fixes #1735

I am seeking feedback regarding the following:

- Should we use a more formal approach to identifying wasm modules?
The basis transcoder uses a naming convention for choosing at runtime which version of the basis module to load based on the runtime support for compressed formats ('basist_pvr.wasm.js for PVR 'basist_astc.wasm.js for ASTC etc etc). Perhaps each module should be assigned a tagging system describing its use (i.e. 'texture', 'compression', 'basis', 'pvr') or path assignment 'texture/compression/basis/pvr'. This may be overly complex.

- What is the best method of supplying the set of basis modules to the engine?
The module configs are currently taken from either window.config.wasmModules or window.PRELOAD_MODULES since that is where the editor and tooling currently stores wasm module details.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/scene/graph-node.js [hunk 1]
```diff
@@ -876,11 +876,9 @@ Object.assign(pc, function () {
         },
 
         _dirtifyLocal: function () {
-            if (!this._dirtyLocal) {
```
Comment (slimbuck) [inline src/scene/graph-node.js]: This check is always done at the call sites. Checking again here is unnecessary.

**Comment by Maksims** [comment]
Is loading basis optional, or it will dramatically increase size of the playcanvas engine, even if basis textures are not used?

**Comment by slimbuck** [comment]
Basis is optional. The basis transcoder module is downloaded asynchronously and used in the web worker only when a basis file is first encountered.

**Comment by Maksims** [comment]
Can it be pre-loaded if developer knows that basis features will be required? That is for the case when none of the basis texture is preloaded, but loaded in async. Is there is a way to set preloading for basis, or triggers its loading?

**Comment by slimbuck** [comment]
Hi @Maksims,

There is currently no way of pre-constructing the basis web worker (a.k.a preloading the wasm module).

Can you suggest how we would expose such an option?

Thanks!

**Comment by slimbuck** [comment]
To give some context regarding the size of wasm modules.

The glue code for all versions of the wasm transcoder is 57k uncompressed and 15k compressed.

Below are the sizes of the wasm binary file for various runtime formats: 
```
format uncomp comp
etc1   64K    35k
pvr    282k   143k
dxt    195k   104k
all    1MB    468k
```

All includes transcoding to all possible runtime formats.

**Comment by slimbuck** [comment]
Also, we have a fork of the Binomial codebase which adds a makefile for generating the required set of binaries (the upstream project generates a single binary). This can be used to create your own version of the basis modules for use in your projects.

The branch which includes the updated makefile is at https://github.com/playcanvas/basis_universal/tree/webgl-transcoder-variants.

**Comment by Maksims** [comment]
Regarding loading module, is it similar to physics case? I believe they fall under same behaviour: engine extensions.
I believe developer should have options to:

1. Use engine without basis functionality (and not be impacted by download of libraries).
2. Just start using engine with basis textures, and basis modules will load automatically.
3. Use engine, and set basis modules to be preloaded, or to be loaded in async later (to keep naked engine lightweight, example would be product customisers, where compressed textures might be loaded only after some initial UI to keep download and loading times as low as possible. and developer want to precisely control assets loading order, this includes engine extensions).

Whatever functionality to enable of such would be.

**Comment by slimbuck** [comment]
Hi @Maksims,

Basis modules differ from ammo modules in two ways. One is that the basis module is downloaded and instantiated in a webworker (when webworkers are supported). This means they can't use the same module logic as ammo (unfortunately). Secondly the graphics device must be present to know which specific basis module to load.

Regarding your points above:
1) This is supported - user must include the basis modules in their project just like ammo.
2) This is also supported - basis module will be loaded by webworker on first encounter of a .basis texture format.
3) This is currently not supported. Preloading would be relatively simple - just instantiate Basis at startup (after the device has been created). However where would the 'preload basis' project setting live? We don't really have a method of storing engine-optional settings. TBD with @willeastcott and @vkalpias et al.

Thanks for the feedback!

File: src/resources/basis.js [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        },
+
+        _handleWorkerResponse: function (message) {
+            var url = message.data.url;
+            var err = message.data.err;
+            var data = message.data.data;
+            var callbacks = this.callbacks[url];
+
+            if (!callbacks) {
+                console.error('internal logical error encountered');
```
Comment (vkalpias) [inline src/resources/basis.js]: Should we be a little more specific with those errors that they are coming from Basis?

File: src/resources/basis.js [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+            var width = basisFile.getImageWidth(0, 0);
+            var height = basisFile.getImageHeight(0, 0);
+            var images = basisFile.getNumImages();
+            var levels = basisFile.getNumLevels(0);
+            var hasAlpha = !!basisFile.getHasAlpha();
+
+            if (!width || !height || !images || !levels) {
+                basisFile.close();
+                basisFile.delete();
+                throw new Error('Invalid image dimensions');
```
Comment (vkalpias) [inline src/resources/basis.js]: It would be helpful to mention that these errors are related to basis

Comment (slimbuck) [reply src/resources/basis.js]: Sure, I'll add the url to the message

File: src/resources/basis.js (hunk 2 above)
Comment (slimbuck) [reply src/resources/basis.js]: Sure, I'll indicate it's the basis transcoder

**Pull request merged**